src/Bicep.Core.UnitTests/Utils/ServiceBuilder.cs

```diff
@@ -0,0 +1,150 @@
+
+        public static ServiceBuilder WithWorkspace(this ServiceBuilder serviceBuilder, IWorkspace workspace)
+            => Register(serviceBuilder, workspace);
+
+        public static ServiceBuilder WithFeatureProvider(this ServiceBuilder serviceBuilder, IFeatureProvider featureProvider)
+            => Register(serviceBuilder, featureProvider);
+
+        public static ServiceBuilder WithNamespaceProvider(this ServiceBuilder serviceBuilder, INamespaceProvider namespaceProvider)
+            => Register(serviceBuilder, namespaceProvider);
+
+        public static ServiceBuilder WithConfigurationManager(this ServiceBuilder serviceBuilder, IConfigurationManager configurationManager)
+            => Register(serviceBuilder, configurationManager);
+
+        public static ServiceBuilder WithApiVersionProvider(this ServiceBuilder serviceBuilder, IApiVersionProvider apiVersionProvider)
+            => Register(serviceBuilder, apiVersionProvider);
+
+        public static ServiceBuilder WithBicepAnalyzer(this ServiceBuilder serviceBuilder, IBicepAnalyzer bicepAnalyzer)
+            => Register(serviceBuilder, bicepAnalyzer);
+
+        public static ServiceBuilder WithTestDefaults(this ServiceBuilder serviceBuilder)
+            => serviceBuilder
+                .WithFileResolver(BicepTestConstants.FileResolver)
+                .WithModuleDispatcher(BicepTestConstants.ModuleDispatcher)
+                .WithWorkspace(new Workspace())
+                .WithFeatureProvider(BicepTestConstants.Features)
+                .WithRegistration(x => x.AddSingleton<INamespaceProvider, DefaultNamespaceProvider>())
+                .WithRegistration(x => x.AddSingleton<IAzResourceTypeLoader, AzResourceTypeLoader>())
+                .WithConfigurationManager(BicepTestConstants.ConfigurationManager)
+                .WithApiVersionProvider(BicepTestConstants.ApiVersionProvider)
+                .WithBicepAnalyzer(BicepTestConstants.LinterAnalyzer);
+
+        public static ServiceBuilder WithAzResources(this ServiceBuilder serviceBuilder, IEnumerable<ResourceTypeComponents> resourceTypes)
+            => serviceBuilder.WithRegistration(x => x.AddSingleton<IAzResourceTypeLoader>(TestTypeHelper.CreateAzResourceTypeLoaderWithTypes(resourceTypes)));
+
+        public static ServiceBuilder WithEmptyAzResources(this ServiceBuilder serviceBuilder)
+            => serviceBuilder.WithAzResources(Enumerable.Empty<ResourceTypeComponents>());
+
+        public static ServiceBuilder WithWorkspaceFiles(this ServiceBuilder serviceBuilder, IReadOnlyDictionary<Uri, string> fileContentsByUri)
+        {
+            var workspace = new Workspace();
+            var sourceFiles = fileContentsByUri.Select(kvp => SourceFileFactory.CreateSourceFile(kvp.Key, kvp.Value));
+            workspace.UpsertSourceFiles(sourceFiles);
+
+            return serviceBuilder.WithWorkspace(workspace);
+        }
+    }
+
+    public class ServiceBuilder
```

Comment:
This is the core logic for this PR. I couldn't come up with a good name for it, so name suggestions very welcome - just would like to keep it short.